### PR TITLE
[batch] Refactor zone quotas code into a zone monitor

### PR DESCRIPTION
**Pull request by jigold**
This PR is the first towards the goal of supporting arbitrary machine types.

I tested both code paths where we create zones from a static list versus if there's over 4000 cores, creating it with east and west regions as well.

I thought about putting the regions and init_zones in the database, but figured that could be a follow up PR. I wasn't sure what the best way to do that was.

### CODE REVIEW

File: batch/batch/driver/zone_monitor.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    async def async_init(self):
+        self.task_manager.ensure_future(retry_long_running(
+            'update_zones_loop',
+            self.update_region_quotas_loop))
+
+    def shutdown(self):
+        self.task_manager.shutdown()
+
+    def zone_weights(self, worker_cores, worker_local_ssd_data_disk, worker_pd_ssd_data_disk_size_gb):
```
Comment (jigold) [inline batch/batch/driver/zone_monitor.py:91]: I refactored this code from what Cotton had as the cores, local ssd, etc are not static in the case of multiple pools and are no longer global variables.

**Comment by jigold** [comment]
FYI: @lgruen

File: batch/batch/driver/main.py [hunk 2]
```diff
@@ -745,6 +746,10 @@ async def on_startup(app):
     log_store = LogStore(BATCH_BUCKET_NAME, WORKER_LOGS_BUCKET_NAME, instance_id, pool, credentials=credentials)
     app['log_store'] = log_store
 
+    zone_monitor = ZoneMonitor(app)
```
Comment (lgruen) [inline batch/batch/driver/main.py:749]: Could we also pass `regions` and `init_zones` here, maybe after adding them to `batch_configuration`?

Comment (jigold) [reply batch/batch/driver/main.py:749]: Exactly, I just wasn't sure the best way to do this. Should it be in the database as a globals field in text that's JSON or should it be passed as part of the deployment.yaml as an environment variable that's also JSON? Perhaps, there's a better way than these options.

Comment (lgruen) [reply batch/batch/driver/main.py:749]: Yet another option might be to move all the environment variables from batch/batch/batch_configuration.py into a single JSON K8s secret, similar to hail/python/hailtop/config/deploy_config.py. But that would be a larger refactoring effort.

File: batch/batch/driver/zone_monitor.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        if self.region_info is None:
+            return self.init_zone_weights
+
+        _zone_weights = []
+        for _, r in self.region_info.items():
```
Comment (danking) [inline batch/batch/driver/zone_monitor.py:96]: use .values() instead.

File: batch/batch/driver/main.py (hunk 2 above)
Comment (danking) [reply batch/batch/driver/main.py:749]: all the config is currently in the db, I'm inclined to stick with that unless we rethink everything. deployment env var seems natural. I don't think we have any strong case currently to change regions or initial zones at run-time.

**Review by danking — CHANGES_REQUESTED** [review]
looks good overall

**Comment by jigold** [comment]
I propose saving the configuration question to a different PR.

**Review by danking — APPROVED** [review]

**Pull request merged**